### ISSUE TITLE
Generator ignoring `include=false` of `tests.toml`

### ISSUE THREAD

**Issue by andrerfcsantos** [exercism/go#2055]
`include=false`  is a directive that can be used in the `tests.toml` file of exercises to tell the tooling that a particular test is not implemented by the track. However, the test generator is ignoring this attribute when generating the tests, when it shouldn't:

> If a track uses a test generator to generate an exercise's test suite, it must use the contents of the tests.toml file to determine which tests to include in the generated test suite.
from [problem-specifications](https://github.com/exercism/problem-specifications#track-test-data)

I ran into this issue when going through the process of updating the tests for `scale-generator` using configlet. This particular exercise had test cases that were replaced using the [reimplements field](https://github.com/exercism/problem-specifications#changing-tests). I told configlet to chose the new version of all the tests, and what configlet did was to update the `tests.toml` with the new tests (by adding them to `tests.toml`), and mark the tests that were reimplemented with `include=false`. However, the generator then proceeded to generate all the tests, regardless of whether they were reimplemented by another test case or not. 

The proposed solution is to follow the recommendation in problem-specifications and make the generator read `tests.toml` and skip generating tests marked with `include=false`.

**Comment by eklatzer** [comment]
Hey, I will try to fix this

**Comment by eklatzer** [comment]
Hey, I came across the following problem. Most of exercise-specific-generators look something like this:
```
func main() {
	t, err := template.New("").Parse(tmpl)
	if err != nil {
		log.Fatal(err)
	}
	var j js
	if err := gen.Gen("luhn", &j, t); err != nil {
		log.Fatal(err)
	}
}

// The JSON structure we expect to be able to unmarshal into
type js struct {
	Cases []struct {
		Description string
		Input       struct {
			Value string
		}
		Expected bool
	}
}
```

All of them define the structure for the `canonical-data.json`. Later when comparing the test-cases to the data from the `tests.toml` every test-case would need the uuid to get the corresponding entry. What is the recommended way of handling this problem to follow guidelines of this project, as there is currently no info about the uuid of each test-case.

**Comment by eklatzer** [comment]
Another question: how should external dependencies be handled? I have created a draft code to play around with the handling of the `.toml` as I have never used one in Go. Therefore I use the external dependency `github.com/pelletier/go-toml/v2`. What is the recommended way of handling this dependency?

```
	testsFilePath := filepath.Join("tests.toml")
	testsFileContent, err := ioutil.ReadFile(testsFilePath)
	if err != nil {
		log.Fatalln("[ERROR] failed to load tests.toml")
	}

	type TOMLFile map[string]struct {
		Description string `toml:"description"`
		Include     bool   `toml:"include"`
		Comment     string `toml:"comment"`
	}

	var tomlContent TOMLFile
	err = toml.Unmarshal(testsFileContent, &tomlContent)
	if err != nil {
		log.Fatalln("[ERROR] failed to unmarshal tests.toml")
	}
	for s, s2 := range tomlContent {
		log.Println(s, ":", s2.Description, "|", s2.Include, "|", s2.Comment)
	}
```

**Comment by eklatzer** [comment]
> Hey, I came across the following problem. Most of exercise-specific-generators look something like this:
> 
> ```
> func main() {
> 	t, err := template.New("").Parse(tmpl)
> 	if err != nil {
> 		log.Fatal(err)
> 	}
> 	var j js
> 	if err := gen.Gen("luhn", &j, t); err != nil {
> 		log.Fatal(err)
> 	}
> }
> 
> // The JSON structure we expect to be able to unmarshal into
> type js struct {
> 	Cases []struct {
> 		Description string
> 		Input       struct {
> 			Value string
> 		}
> 		Expected bool
> 	}
> }
> ```
> 
> All of them define the structure for the `canonical-data.json`. Later when comparing the test-cases to the data from the `tests.toml` every test-case would need the uuid to get the corresponding entry. What is the recommended way of handling this problem to follow guidelines of this project, as there is currently no info about the uuid of each test-case.

Regarding this issue we could use something like this: Instead of passing a `interface{}` as storage for the JSON data to `func Gen`, we could create our own interface:
```
type TestData interface {
	GetUUIDs() []string
	ExcludedTests([]string)
}
```
Then we have to adjust `func Gen` something like this (pseudo-code):
```
includedTests := j.GetUUIDs()
testsToExclude := checkTOML(includedTests)
j.ExcludedTests(testsToExclude)
```

**Comment by junedev** [comment]
@eklatzer I don't think changing the signature of `Gen` (and with it a lot of code in all the exercises) is necessary to solve the issue at hand. I would recommend the following approach where we modify the canonical-data.json data to out liking before passing it on.
1. Inside the `Gen` function, parse the toml to get the list of uuids to exclude.
2. Unmarshal the `canonical-data.json` file content (`jSrc`) into a `map[string]interface{}` or something similar that is rather generic.
3. Look for the cases key and assert its value to be `[]map[string]interface{}`. Now we have a list of cases.
4. Go through the list of cases. If the item has a uuid and that uuid is in the uuids to exclude, remove the item from the list.
5. If the item has another cases key, do step 3-5 for this "cases" entry as well (recursively).
6. Once all unwanted tests were thrown out, marshal back into bytes. From here, the rest of the code can be executed as usual just with the modified canonical data bytes as input for the unmarshal.

Probably there can be even better data structures for 2 and 3 so there are less type assertions necessary.

Re third party libraries: It is totally fine to use them if necessary. You can do a `go mod init` in the `gen` folder for the dependency management. You can see an example for a tool with external dependencies in the `gomod-sync` folder. Please use the same pattern for the module name and the same go version in the go.mod file as you can see there.

**Comment by junedev** [comment]
@AbhilashJN I thought the only reason the generator needs to be run with that environment variable is that there was no go mod file in the directory. If this wasn't the only reason, do you know why the env is needed in the first place?

**Comment by junedev** [comment]
@AbhilashJN Yes, that makes sense. I read that it is possible to use the replace directive for local imports (https://thewebivore.com/using-replace-in-go-mod-to-point-to-your-local-module/) but that would mean to add this replace directive in the go.mod file of each exercise. 
The better solution seems to be to use the new workspace mode that was introduced with Go 1.18: https://go.dev/doc/tutorial/workspaces From the description in the tutorial it sounds like this is what we want (1 local folder that contains multiple modules where one can reference the other). Going this route would mean we require every contributor to have Go 1.18 installed so I would like to wait with this a bit longer so we don't exclude people from working on e.g. exercise updates.

Until Go 1.18 is more widespread I would suggest to either do the manual toml parsing (the format is quite stable so it should be fine for now) or to copy and paste the toml parser module code into a subfolder of the generator.
Then at some point (at latest when Go 1.18 goes out of maintenance) we can set up the workspace mode and use the toml parser module via the normal dependency mgmt.
